### ISSUE TITLE
LoRA Training notes

### ISSUE THREAD

**Issue by vladmandic** [vladmandic/onetrain#9]
I've ran more than few training sessions for SDXL to get to best practices.

## Dataset

This means the most as single low image can ruin the quality of LoRA and its better to have smaller number of good images than large number of medium quality images. At the end, I've settled down on testing using same uniform dataset with 12 images

*Example* raw input dataset: <https://github.com/vladmandic/onetrain/issues/6>

Onetrain was additionally configured to
- Extract and create face close-ups from input dataset which resulted in additional 10 images being auto-created so total was 22 images.
- Resize images
- Optimize buckets

Note that test with 10 images only showed that is sufficient for LoRA, but more images makes LoRA more flexible.

## Base model

This has far bigger impact than I originally thought and for a different reason: when using a fine-tuned model, almost every finetune is numerically overly-polished meaning training on such models will result in highly vaxy-looking LoRA without any details in textures.

This means you can use fine-tuned model for generate, but for LoRA training you want to use as-raw-as-possible model.
To avoid testing 20+ models, I've settled down using original SDXL-base as base model for training which resulted in massive increase in quality.

For reference, link to SDXL base model: <https://civitai.com/models/101055/sd-xl>

*Example* using fine-tuned model: <https://github.com/vladmandic/onetrain/issues/8#issuecomment-2676268642>

## Caption

As expected, first entry in caption file needs to be concept name itself (trigger word).
For the rest, i've used `wdtagger` and `cogflorence` so ending caption ended up being pretty long, but it does allow for easy prompting when using LoRA during generate:

*Example* caption: 
> _li, 1girl, solo, sensitive, long hair, realistic, window, looking at viewer, brown hair, swimsuit, parted lips, ass, indoors, bikini, photorealistic, from side, lips, asian, portrait of woman standing by window. She wears black, strappy bralette and high-waisted shorts. Her long, straight brown hair flows down her back. She adjusts her hair with one hand, her gaze directed away from the camera. window behind her has white frame and golden handle. background is light-colored wall with visible view of greenery outside. image style is contemporary, emphasizing the subject's form and the natural setting

*TODO*: test training with TE+UNET typical training is UNET-only

At the end, processed training dataset with captions: <https://github.com/vladmandic/onetrain/issues/7>

## Optimizer/Schedule

I'm using `COSINE` scheduler as its math properties result in stable learning rate.
Using `constant` scheduler is possible if you want to manually tweak every LR param manually, but I see no need 

Equally, I'm using `PRODIGY` optimizer - which means that manually set LR is ignored: prodigy main advantage is that it auto-adjust learning rate as needed. You can accelerate or slow down adjustments using `d_coef` (0.5 would make learning-rate adjustments twice as slow meaning it will be less aggressive and 2.0 would make it twice as fast adjustments making it more aggressive). I've settled down at value of 0.8-0.9, but if you want to have faster training while sacrificing just a little bit of details, you can accelerate it to 1.5+

Another param that does have an impact on auto-adjusting optimizers is warmup_steps as it determines how many training steps will be performed to purely determine LR and those steps are discarded. I'm using 100 warmup steps, but if you want to speed up training while resulting in very slightly worse results, that can be reduced to 25-50.

Optimizer that produces very similar results is `ADAFACTOR`.
I did test non-dynamic optimizers but saw no increase in quality and it just adds complexity due to needing to manually managing all learning rate related params.

Additionally, quantized optimizers such as `ADAMW8BIT` only exist so training can be performed with very-low VRAM as it happens in 8-bit.
If you have sufficient VRAM, dont use it - as simple as that.
Also, using any 8-bit-or-lower quantized model or training requires quantization support using `bitsandbytes` which is not installed by default.

*Example* calculated LR vs steps when using different d_coef values
![Image](https://github.com/user-attachments/assets/23766f0a-43c5-4402-b7fb-70f3c9420198)

*Example* using `ADAFACTOR` optimizer: <https://github.com/vladmandic/onetrain/issues/8#issuecomment-2676273165>

## Precision

BFLoat16 is recommended during generate as its simply less prone to overflows resulting in black images.
But for training itself, BFloat16 does not have sufficient precision. So always train in Float16 (given that FP32 requires twice as much resources/time, it is not recommended)

## Batch/Accumulation

Using batch size and/or gradient accumulation allows training to learn from multiple images at once.
Which is great if you're trying to learn general concepts. But for high-quality person/face training, this results in distortions and or blurry results as there almost never multiple images in input dataset that can be overlaid without creating a blur or an effect with bad anatomy.
Which means that training is not going to utilize GPU fully, but if quality if imperative only solution is to use batch_size 1.

*Example* with batch-size=1: <https://github.com/vladmandic/onetrain/issues/8#issuecomment-26762686426>
*Example* with batch-size=2: <https://github.com/vladmandic/onetrain/issues/8#issue-2870896286>

## Rank/Alpha

Rank (or dim) simply increases LoRA size which means more room to learn. But learning based on dataset of 10-20 and high rank is pointless (confirmed via experiments). As a result, I've settled down on rank=16/alpha=16 which results in ~95MB LoRA file.

## Epochs/Steps

I'm running very long training with 100 epochs, but checking progress every 10 epochs.
100 epochs at batch 1 with 22 input images means training is going to run for 22/1*100=2200 steps.
If you want to have fully predictable training length, you can use max steps instead of max epochs as total steps depends on number of input images and batch size.

What are the results?
- Epoch 0: using test prompt only, no LoRA impact as we're inside warmup phase for first 100 steps anyhow
- Epoch 10-20: prodigy uses very high learning rate so it very quickly starts converging. LoRA at epoch=20 is already usable and this only takes 15min of training time. What's missing is fine details.
- Epoch 30-40: scheduler re-introduces noise which results in visually worse LoRA at this time. However, this is needed to break the local minimum so further training can add details.
- Epochs 50+: scheduler continues with much lower learning rate and gradually adds details over time. With auto-adjusting optimizer, there is no cutoff when LoRA becomes over-baked as it simply continues using lower and lower LR. With classic scheduler, long training woulddefinitely result in over-training.

Total training time for 2200 steps is ~75min and unnecessary, the question what is "good enough": 
- 15-20 epochs for quick training in 15min
- 50-60 epochs for detailed training in 35min

*Examples*: 
- <https://github.com/vladmandic/onetrain/issues/8#issuecomment-2676274944>
- <https://github.com/vladmandic/onetrain/issues/8#issuecomment-2676275336>

## Configuration

Base: 
[config.json](https://github.com/user-attachments/files/18923766/config.json)

Overrides:
[lora-sdxl.json](https://github.com/user-attachments/files/18923765/lora-sdxl.json)

## Other

To allow LoRA to break forementioned local minimum and continue training fine details instead of getting stuck, I'm using relatively high value of `weight_decay=0.01` which means 1% of learned weights are re-noised in every step.
Using lower weight decay can reduce effect of interim LoRA degradations and produce faster end-result.
